pkg/controller/elasticsearch/pdb/reconcile.go

```diff
@@ -7,107 +7,165 @@ package pdb
+	if template == nil {
+		template = &commonv1alpha1.PodDisruptionBudgetTemplate{}
+	}
+
+	expected := v1beta1.PodDisruptionBudget{
+		ObjectMeta: template.ObjectMeta,
+	}
+
+	// inherit user-provided ObjectMeta, but set our own name & namespace
+	expected.Name = name.DefaultPodDisruptionBudget(es.Name)
+	expected.Namespace = es.Namespace
+	// and append our labels
+	expected.Labels = defaults.SetDefaultLabels(expected.Labels, label.NewLabels(k8s.ExtractNamespacedName(&es)))
+	// set owner reference for deletion upon ES resource deletion
+	if err := controllerutil.SetControllerReference(&es, &expected, scheme); err != nil {
+		return nil, err
+	}
+
+	if template.Spec.Selector != nil || template.Spec.MaxUnavailable != nil || template.Spec.MinAvailable != nil {
+		// use the user-defined spec
+		expected.Spec = template.Spec
+		return &expected, nil
+	}
+
+	// set our default spec
+	var err error
+	if expected.Spec, err = buildPDBSpec(es, statefulSets); err != nil {
+		return nil, err
+	}
+
+	return &expected, nil
+}
+
+// buildPDBSpec returns a PDBSpec computed from the current StatefulSets,
+// considering the cluster health and topology.
+func buildPDBSpec(es v1alpha1.Elasticsearch, statefulSets sset.StatefulSetList) (v1beta1.PodDisruptionBudgetSpec, error) {
+	// compute MinAvailable based on the maximum number of Pods we're supposed to have
+	nodeCount := statefulSets.ExpectedPodCount()
+	// maybe allow some Pods to be disrupted
+	minAvailable := nodeCount - allowedDisruptions(es, statefulSets)
+
+	minAvailableIntStr := intstr.IntOrString{Type: intstr.Int, IntVal: minAvailable}
+
+	return v1beta1.PodDisruptionBudgetSpec{
+		// match all pods for this cluster
+		Selector: &metav1.LabelSelector{
 			MatchLabels: map[string]string{
 				label.ClusterNameLabelName: es.Name,
 			},
-		}
-	}
-
-	var reconciled v1beta1.PodDisruptionBudget
-	return reconciler.ReconcileResource(reconciler.Params{
-		Client:     c,
-		Scheme:     scheme,
-		Owner:      &es,
-		Expected:   &expected,
-		Reconciled: &reconciled,
-		NeedsUpdate: func() bool {
-			for k, v := range expected.Labels {
-				if rv, ok := reconciled.Labels[k]; !ok || rv != v {
-					return true
-				}
-			}
-
-			return !reflect.DeepEqual(expected.Spec, reconciled.Spec)
 		},
-		UpdateReconciled: func() {
-			for k, v := range expected.Labels {
-				reconciled.Labels[k] = v
-			}
-			reconciled.Spec = expected.Spec
-		},
-	})
+		MinAvailable: &minAvailableIntStr,
+		// MaxUnavailable can only be used if the selector matches a builtin controller selector
+		// (eg. Deployments, StatefulSets, etc.). We cannot use it with our own cluster-name selector.
+		MaxUnavailable: nil,
+	}, nil
+}
+
+// allowedDisruptions returns the number of Pods that we allow to be disrupted while keeping the cluster healthy.
+func allowedDisruptions(es v1alpha1.Elasticsearch, actualSsets sset.StatefulSetList) int32 {
+	if es.Status.Health != v1alpha1.ElasticsearchGreenHealth {
```

Comment:
Q: will this get triggered if ES changes state to yellow on it's own, ie. without any changes visible from k8s?

Reply:
Yes! We [trigger a reconciliation when the observed ES health changes](https://github.com/elastic/cloud-on-k8s/blob/260aff6c5d7026e58315a309f81b973b26d17f50/pkg/controller/elasticsearch/observer/watch.go#L32).